plugins/ingest/src/main/java/org/elasticsearch/ingest/IngestDocument.java

```diff
@@ -339,12 +375,51 @@ private void setFieldValue(String path, Object value, boolean append) {
             if (index < 0 || index >= list.size()) {
                 throw new IllegalArgumentException("[" + index + "] is out of bounds for array with length [" + list.size() + "] as part of path [" + path + "]");
             }
+            if (append) {
+                Object object = list.get(index);
+                List<Object> newList = appendValues(path, object, value);
+                if (newList != object) {
+                    list.set(index, newList);
+                }
+                return;
+            }
             list.set(index, value);
         } else {
             throw new IllegalArgumentException("cannot set [" + leafKey + "] with parent object of type [" + context.getClass().getName() + "] as part of path [" + path + "]");
         }
     }
 
+    @SuppressWarnings("unchecked")
+    private static List<Object> appendValues(String path, Object maybeList, Object value) {
+        if (maybeList == null) {
+            throw new IllegalArgumentException("cannot append value to existing null element at [" + path + "]");
+        }
+        if (maybeList instanceof Map) {
```

Comment:
hey @mvg question: I think we should create a list of maps in this case too? thinking more about it, not sure this special case that throws exception makes sense anymore.